### ISSUE TITLE
Excessive constraints generated by lemma applications

### ISSUE THREAD

**Issue by rybla** [ucsd-progsys/liquidhaskell#1944]
Consider this schema:
```haskell
{-@ measure r :: Int -> Bool @-}

{-@ assume lem :: x:{Int | r x} -> {r (x + 1)} @-}
lem :: Int -> ()
lem _ = ()

with :: a -> b -> a
with a _ = a

{-@
test :: x:{Int | r x} -> {r (x + N)}
@-}
test :: Int -> ()
test x = 
  lem x `with`
  lem (x + 1) `with`
  ...
  lem (x + (N - 1))
```
where `N` is some constant integer greater than 1.

I noticed that the number of constraints generated by this simple schema seems excessive. Here is some data, where "lemma applications" corresponds to the `N` above:

[Lemma application constraint generation.pdf](https://github.com/ucsd-progsys/liquidhaskell/files/8104884/Lemma.application.constraint.generation.pdf)

Visual summary:
![Lemma application constraint generation](https://user-images.githubusercontent.com/7615148/154863505-1e5c308c-5e57-4787-ad80-0923628eed1a.png)

The data covers a few variants:

**"with TH"** indicates that the imports
```
import Language.Haskell.TH
import Language.Haskell.TH.Quote
import Language.Haskell.TH.Syntax
```
were included at the beginning of the file.

**"without TH"** indicates that those imports were not included at the beginning of the file.

**"monomorphic with"** indicates that `with` was retyped to be `with :: a -> a -> a`.

So, why are so many constraints being generated (20,000+ for just 10 applications of the simple lemma!) and why does it depend on importing the TH modules (which are not used in the code) and whether `with` is monomorphic or not?

**Comment by facundominguez** [comment]
I can't reproduce very well with `develop`. I only get near 700 and 800 constraints at most in all cases. Are you testing with the latest `liquidhaskell` on hackage maybe?

Also, `Proof` is not in scope in the example. I replaced it with `()`.

If I use
```
with :: () -> () -> ()
```
then I get only 19 constraints checked.

Polymorphism is known to cause some explosions [elsewhere](https://github.com/ucsd-progsys/liquid-fixpoint/issues/482).

**Comment by ranjitjhala** [comment]
Good point -- Henry, can you let us know what git commit you're using?
Thanks!

**Comment by rybla** [comment]
Yeah I think its the polymorphism that is responsible for the constraint explosions.

**Comment by rybla** [comment]
I am using `liquidhaskell == 0.8.10.7`

**Comment by rybla** [comment]
Any idea about the TH imports?
